src/xmpp/xmpp-im/xmpp_mamtask.cpp

```diff
@@ -0,0 +1,236 @@
+    }
+
+    query.appendChild(x.toXml(t->doc()));
+    query.appendChild(rsm.makeQueryElement(t->doc()));
+    iq.appendChild(query);
+    t->send(iq);
+}
+
+void MAMTask::Private::getArchiveMetadata(MAMTask *t)
+{
+    // Craft a query to get the first and last messages in an archive
+    QDomElement iq       = createIQ(t->doc(), QLatin1String("get"), QLatin1String(), t->id());
+    QDomElement metadata = emptyTag(t->doc(), QLatin1String("metadata"));
+    metadata.setAttribute(QLatin1String("xmlns"), XMPP_MAM_NAMESPACE);
+    iq.appendChild(metadata);
+    iq.appendChild(makeMAMFilter().toXml(t->doc()));
+
+    t->send(iq);
+}
+
+MAMTask::MAMTask(Task *parent) : Task(parent) { d = new Private; }
+MAMTask::~MAMTask() { delete d; }
+
+const QList<QDomElement> &MAMTask::archive() const { return d->archive; }
+
+// Filter by time range
+void MAMTask::get(const Jid &j, const QDateTime &from = {}, const QDateTime &to = {},
+                  const bool allowMUCArchives = true, int mamPageSize = 10, int mamMaxMessages = 100,
+                  bool flipPages = true, bool backwards = true)
+{
+    d->archive         = {};
+    d->messagesFetched = 0;
+    d->metadataFetched = false;
+
+    d->j                = j;
+    d->from             = from;
+    d->to               = to;
+    d->allowMUCArchives = allowMUCArchives;
+    d->mamPageSize      = mamPageSize;
+    d->mamMaxMessages   = mamMaxMessages;
+    d->flipPages        = flipPages;
+    d->backwards        = backwards;
+}
+
+// Filter by id range
+void MAMTask::get(const Jid &j, const QString &from_id = {}, const QString &to_id = {},
+                  const bool allowMUCArchives = true, int mamPageSize = 10, int mamMaxMessages = 100,
+                  bool flipPages = true, bool backwards = true)
+{
+    d->archive         = {};
+    d->messagesFetched = 0;
+    d->metadataFetched = false;
+
+    d->j                = j;
+    d->from_id          = from_id;
+    d->to_id            = to_id;
+    d->allowMUCArchives = allowMUCArchives;
+    d->mamPageSize      = mamPageSize;
+    d->mamMaxMessages   = mamMaxMessages;
+    d->flipPages        = flipPages;
+    d->backwards        = backwards;
+}
+
+void MAMTask::onGo() { d->getArchiveMetadata(this); }
+
+bool MAMTask::take(const QDomElement &x)
+{
+    if (d->metadataFetched) {
+        if (iqVerify(x, QString(), id()))
+            return false;
+
+        // TODO: only save messages directed at the correct resource?
+        d->archive.append(x);
+        d->lastArchiveID   = x.attribute(QLatin1String("id"));
+        d->messagesFetched = d->messagesFetched + 1;
+
+        // Check if we are done
+        if (x.attribute(QLatin1String("id")) == d->lastID || d->messagesFetched >= d->mamMaxMessages) {
+            setSuccess();
+        } else if (d->messagesFetched % d->mamPageSize == 0) {
+            d->getPage(this);
+        }
+        // TODO: handle server not sending all MAM stuff gracefully (such as handling 0 messagesFetched)
```

Comment:
Error handling

Reply:
@Ri0n how does iris handle tasks that hang too long? If the server sends a malformed response, the task will simply stop running, and the messages that were received properly can be accessed with the `archive()` getter.

Reply:
there is 2 minutes timeout